### PR TITLE
Support for Hardhat Network

### LINKED ISSUE #1024: [hardhat] Incorrect transaction gas limit
### Environment information

* `brownie` Version: 48d3f182747dc7807c4d6149024b519dd7215061
* `hardhat` Version: 2.1.3-dtt.0 sha512-QWI3LbImf5vyshQzC2nGxzkgtMHsK5BovcWUMgNl3f/0QGCtVRLhTvkA6/FK0f8XhDkFCOHLsJmyRrt4Ao9grg==
* `solc` Version: 0.8.3
* Python Version: 3.9.2
* OS: osx and linux

### What was wrong?

I'm testing out swapping ganache-cli for hardhat@dtt on the [argobytes](https://github.com/SatoshiAndKin/argobytes-contracts) repo.

I then ran `hardhat node --fork https://mynode` and then `brownie test --network mainnet-fork`

Some of my tests failed with `Transaction gas limit is 12475452 and exceeds block gas limit of 9500000`. This is the mainnet gas limit and not hardhat's limit.

### Workaround

Add this to your hardhat config:
```
module.exports = {
  networks: {
    hardhat: {
      blockGasLimit: 14000000
    }
  }
};
```

### How can it be fixed?

Ben Hauser, [Mar 30, 2021 at 1:38:50 PM]:
brownie chooses the gas limit in dev mode by looking at the block gas limit on the last block
so prolly just need to re-evaluate how that works

### LINKED ISSUE #1025: [hardhat] unlocked account is missing
### Environment information

* `brownie` Version: 48d3f182747dc7807c4d6149024b519dd7215061
* `hardhat` Version: 2.1.3-dtt.0 sha512-QWI3LbImf5vyshQzC2nGxzkgtMHsK5BovcWUMgNl3f/0QGCtVRLhTvkA6/FK0f8XhDkFCOHLsJmyRrt4Ao9grg==
* `solc` Version: 0.8.3
* Python Version: 3.9.2
* OS: osx and linux

### What was wrong?

I'm testing out swapping ganache-cli for hardhat@dtt on the [argobytes](https://github.com/SatoshiAndKin/argobytes-contracts) repo.

My tests have this fixture:
```
@pytest.fixture(scope="function")
def unlocked_binance():
    return accounts.at("0x85b931A32a0725Be14285B66f1a22178c672d69B", force=True)
```

However, when I try to use this account, I get a ValueError:
```
>       token.transfer(to, amount, {"from": unlocked_binance})
E       ValueError: unknown account 0x85b931a32a0725be14285b66f1a22178c672d69b
```

### How can it be fixed?

I think hardhat does account unlocking differently than ganache. I haven't looked into it yet.

### LINKED ISSUE #1026: [hardhat] network config adds incompatible flags
### Environment information

* `brownie` Version: 48d3f182747dc7807c4d6149024b519dd7215061
* `hardhat` Version: 2.1.3-dtt.0 sha512-QWI3LbImf5vyshQzC2nGxzkgtMHsK5BovcWUMgNl3f/0QGCtVRLhTvkA6/FK0f8XhDkFCOHLsJmyRrt4Ao9grg==
* `solc` Version: 0.8.3
* Python Version: 3.9.2
* OS: osx and linux

### What was wrong?

I'm testing out swapping ganache-cli for hardhat@dtt on the [argobytes](https://github.com/SatoshiAndKin/argobytes-contracts) repo.

I then ran `brownie test --network mainnet-fork`, but it fails because it tries to use this command:
```
Launching 'hardhat node --fork https://yournode --port 8575 --chainId 1 --hardfork istanbul'...
```

Neither --chainId, nor --hardfork are flags in hardhat. You can set them, but instead of with flags, they are set in `hardhat.config.js` 

### Workaround

In a separate terminal, run `hardhat node --fork https://yournode`

### How can it be fixed?

Either don't pass these flags, or add them to hardhat.

### LINKED ISSUE #1027: [hardhat] brownie reverts context manager does not catch reverts
### Environment information

* `brownie` Version: 48d3f182747dc7807c4d6149024b519dd7215061
* `hardhat` Version: 2.1.3-dtt.0 sha512-QWI3LbImf5vyshQzC2nGxzkgtMHsK5BovcWUMgNl3f/0QGCtVRLhTvkA6/FK0f8XhDkFCOHLsJmyRrt4Ao9grg==
* `solc` Version: 0.8.3
* Python Version: 3.9.2
* OS: osx and linux

### What was wrong?

I'm testing out swapping ganache-cli for hardhat@dtt on the [argobytes](https://github.com/SatoshiAndKin/argobytes-contracts) repo.

One of my tests expects a revert, but it is not being caught.

```
        with brownie.reverts(""):
>           argobytes_proxy_clone.execute(action, {"from": accounts[1]})
E           ValueError: Transaction reverted without a reason

tests/test_argobytes_clone.py:26: ValueError
```

### How can it be fixed?

My guess is that ganache errors were not ValueErrors like hardhat's are. Either change hardhat errors to match ganache, or update the context manager to catch ValueErrors.

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
Add preliminary support for [Hardhat Network](https://hardhat.org/hardhat-network/) as a testing environment.

To test this, you must currently install the `hardhat@dtt` package. The required features for this integration will not merge to the main hardhat repo until after the Berlin fork, in the mean time this PR should also be considered a WIP.

To test, create a Brownie project and then install hardhat inside it:

```bash
brownie init
npm install --save-dev hardhat@dtt
touch hardhat.config.js
```

You can now use hardhat as a dev environment with `--network hardhat` or hardhat fork mode using `--network hardhat-fork`.

### How I did it
Most of the groundwork for this was laid in previous PRs (#997, #998). The integration is made possible via a new backend module and a hardhat-specific middleware.

### How to verify it
Try it! And please report any issues you encounter.

### TODO
Before this merges, the test suite should be modified to target both hardhat and ganache. We also need to start use this in our daily lives and smoke out any weird edge cases or other oddities.

Closes #1024
Closes #1025
Closes #1026
Closes #1027

### CODE REVIEW

**Comment by iamdefinitelyahuman** [comment]
Closes #1024 #1025 #1026 #1027

**Comment by WyseNynja** [comment]
Looks great! I think this covers everything I ran into.

File: brownie/network/middlewares/hardhat.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        if method == "eth_sendTransaction" and "error" in result:
+            message = result["error"]["message"]
+            if not message.startswith("VM Exception"):
+                return result
+            txid = self.w3.eth.getBlock("latest")["transactions"][0]
+            data: Dict = {}
+            result["error"]["data"] = {txid.hex(): data}
+            message = message.split(": ", maxsplit=1)[-1]
+            if message.startswith("revert"):
+                data.update(error="revert", reason=message[7:])
```
Comment (banteg) [inline brownie/network/middlewares/hardhat.py:29]:
```suggestion
            if not (message.startswith("VM Exception") or message == "Transaction reverted without a reason"):
                return result
            txid = self.w3.eth.getBlock("latest")["transactions"][0]
            data: Dict = {}
            result["error"]["data"] = {txid.hex(): data}
            if message.startswith("revert"):
                data.update(error="revert", reason=message[7:])
            elif message == "Transaction reverted without a reason":
                data.update(error="revert", reason=None)
```

File: brownie/network/rpc/hardhat.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/python3
+
+import sys
+import warnings
```
Comment (banteg) [inline brownie/network/rpc/hardhat.py:4]:
```suggestion
import warnings
from pathlib import Path
```

File: brownie/network/rpc/hardhat.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            cmd_list.extend([CLI_FLAGS[key], str(value)])
+        except KeyError:
+            warnings.warn(
+                f"Ignoring invalid commandline setting for hardhat: "
+                f'"{key}" with value "{value}".',
+                InvalidArgumentWarning,
+            )
+    print(f"\nLaunching '{' '.join(cmd_list)}'...")
+    out = DEVNULL if sys.platform == "win32" else PIPE
+
```
Comment (banteg) [inline brownie/network/rpc/hardhat.py:39]:
```suggestion

    hardhat_config = Path("hardhat.config.js")
    if not hardhat_config.exists():
        hardhat_config.touch()
```

**Review by banteg — COMMENTED** [review]
- #1024  **Fixed** `--network hardhat-fork` doesn't work yet, all transactions fail with `ValueError: Transaction gas limit is 12475600 and exceeds block gas limit of 9500000`. Fixed by overriding the limit when creating a hardhat config.
- #1025 **Tested**, impersonation works well.
- #1026 **Tested**, one caveat. Spawning hardhat fails because it seems to go into an interactive prompt suggesting creating a project. Brownie should create `hardhat.config.js` prior to spawning hardhat. Fixed in suggestion.
- #1027 **Fixed** `brownie.reverts` doesn't work properly, getting `ValueError: Transaction reverted without a reason` instead. Fixed from Brownie side. The issue also mentions a fix via hardhat config.

File: brownie/network/rpc/hardhat.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from typing import Dict, List, Optional
+
+import psutil
+from requests.exceptions import ConnectionError as RequestsConnectionError
+
+from brownie.exceptions import InvalidArgumentWarning, RPCRequestError
+from brownie.network.web3 import web3
+
+CLI_FLAGS = {"port": "--port", "fork": "--fork", "fork_block": "--fork-block-number"}
+
```
Comment (banteg) [inline brownie/network/rpc/hardhat.py:15]:
```suggestion

DEFAULT_HARDHAT_CONFIG = """\
module.exports = {
  networks: {
    hardhat: {
        blockGasLimit: 15000000
    }
  }
}
"""
```

**Review by banteg — APPROVED** [review]
Fixed mentioned issues in suggestions. Looks good after that.

**Comment by banteg** [comment]
There are still problems, cont'd in #1044

**Comment by iamdefinitelyahuman** [comment]
Need to expose https://hardhat.org/guides/mainnet-forking.html#pinning-a-block and think about some sensible defaults for it

**Comment by iamdefinitelyahuman** [comment]
Also lots of potentially interesting stuff here: https://hardhat.org/hardhat-network/#mining-modes

**Comment by banteg** [comment]
Seems like one quality of life improvement would be an ability to inject Hardhat config via RPC.

**Comment by banteg** [comment]
Pls cherry pick this too https://github.com/eth-brownie/brownie/pull/1044/commits/a06a35f00c0fbcd2f141201a6ca20d3e905ff426

**Comment by PatrickAlphaC** [comment]
This is beautiful. 

Just ran a few tests.... they are so FAST!

**Comment by iamdefinitelyahuman** [comment]
Some weird inconsistencies between ganache and hardhat with how time is handled.

Implementing https://github.com/eth-brownie/brownie/issues/709 alongside this could help

**Comment by banteg** [comment]
`hardhat@dtt.1` additionally prefixes error messages with "Error: ", not sure if the change is intended.

A fix could look like this, but ideally we need to use error codes to not depend on parsing error messages.
```python
message = result["error"]["message"].replace('Error: ', '')  # fix for hardhat@dtt.1
```

**Comment by PatrickAlphaC** [comment]
Looks like Yearn pulled this into their master branch. Is it still waiting fixes here? https://github.com/yearn/yearn-vaults/pull/268

**Comment by PatrickAlphaC** [comment]
Bump, what does this need to move forward?

**Comment by iamdefinitelyahuman** [comment]
was waiting for hardhat official release - we got that a few days ago, now i guess the blocker is probably me?  really just needs another round of testing and rechecking that everything works as expected, updates to docs...  i'm on vacation at the moment tho so moving slowly.

**Comment by PatrickAlphaC** [comment]
Let me know how I can help

**Comment by pandadefi** [comment]
How can I help?

**Comment by iamdefinitelyahuman** [comment]
This is _mostly_ working. The test suite fails from 3 broad issues:

1. Hardhat returns 21,001 instead of 21,000 when estimating gas on a plain tx, so a bunch of places in brownie's test suite break due to an assumption of 21,000.
2. Hardhat sets a default of 10,000 ether per account, which breaks assumptions around 100 ether per account (ganache's default).
3. The lack of a global hardhat install breaks any tests that involve sub-processes launching in different folders, Mostly this affects the tests of the pytest plugin.

I think this is far enough along to launch it as a sort-of beta integration.  As we use it in prod we'll find issues and shortcomings and I'm sure in a month or 2 it will look better than it does today.  But for now..  lfgggggg

**Comment by iamdefinitelyahuman** [comment]
Docs updates will come in another PR, I need to do some updates around the EIP1559 functionality as well prior to pushing out a `1.16.0` release.

**Review by banteg — APPROVED** [review]

**Comment by sambacha** [comment]
> @iamdefinitelyahuman: 3. The lack of a global hardhat install breaks any tests that involve sub-processes launching in different folders, Mostly this affects the tests of the pytest plugin.
> 

by global hardhat install, do you mean `hh`? see: https://www.npmjs.com/package/hardhat-shorthand
